internal/os/filesystem/api.go

```diff
@@ -0,0 +1,50 @@
+package filesystem
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+)
+
+// Implements the Filesystem OS API calls. All code here should be very simple
+// pass-through to the OS APIs. Any logic around the APIs should go in
+// internal/server/filesystem/server.go so that logic can be easily unit-tested
+// without requiring specific OS environments.
+
+type APIImplementor struct{}
+
+func New() APIImplementor {
+	return APIImplementor{}
+}
+
+func (APIImplementor) PathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return true, err
+}
+
+func (APIImplementor) Mkdir(path string) error {
+	err := os.MkdirAll(path, 0755)
+	return err
+}
+
+func (APIImplementor) Rmdir(path string) error {
+	err := os.RemoveAll(path)
+	return err
+}
+
+func (APIImplementor) LinkPath(tgt string, src string) error {
+	var err error
+	if runtime.GOOS == "windows" {
+		_, err = exec.Command("cmd", "/c", "mklink", "/D", tgt, src).CombinedOutput()
```

Comment:
Wouldn't we want to use async mode in these cases - link I am guessing is not time consuming so not exactly relevant. But from the code flow from the top of the server, shouldn't we have a context which can be passed on to server to cancel the operation, which should be passed up to here and then we can use exec.Start/exec.Wait and exec.Process.Kill to achieve that pattern ? WDYT ? Perhaps we don't need to incur the complexity of this as of now, but just wanted to get an idea of what the direction would be for context/cancellation in client<->server for csi-proxy in general ?

Reply:
Yes we should definitely use the ctx for more the more complex/time consuming operations like formatting a partition, etc as part of the disk APIs. The ctx is passed through to the server layer and ignored for now as discussed [above](https://github.com/kubernetes-csi/csi-proxy/pull/11#discussion_r339925937)